haystack/backends/solr_backend.py

```diff
@@ -389,13 +393,38 @@ def _process_results(self, raw_results, highlight=False, result_class=None, dist
                                                         facets[key][facet_field][1::2]))
 
         if self.include_spelling and hasattr(raw_results, 'spellcheck'):
-            # Solr 5+ changed the JSON response format so the suggestions will be key-value mapped rather
-            # than simply paired elements in a list, which is a nice improvement but incompatible with
-            # Solr 4: https://issues.apache.org/jira/browse/SOLR-3029
-            if len(raw_results.spellcheck.get('collations', [])):
-                spelling_suggestion = raw_results.spellcheck['collations'][-1]
-            elif len(raw_results.spellcheck.get('suggestions', [])):
-                spelling_suggestion = raw_results.spellcheck['suggestions'][-1]
+            # There are many different formats for Legacy, 6.4, and 6.5
+            # e.g. https://issues.apache.org/jira/browse/SOLR-3029
+            collations = raw_results.spellcheck.get('collations', [])
+            suggestions = raw_results.spellcheck.get('suggestions', [])
+            if len(collations):
+                #Handle sol6.5 collation format
+                if isinstance(collations, dict):
+                    spelling_suggestions= [col['collationQuery'] for col in collations.values()]  #aggregate for future use in multi suggestion response
+                #Legacy Legacy & 6.4 handling
+                else:
+                    if isinstance(collations[1], dict):  #Solr6.4
+                        spelling_suggestions = [item["collationQuery"] for item in collations if isinstance(item,dict)]  #aggregate for future use in multi suggestion response
+                    else:  #Legacy Solr format
+                        spelling_suggestions=collations[-1]
+
+                spelling_suggestion = spelling_suggestions[-1]  #Keep current method of returning single value
+            elif len(suggestions):
+                #Handle sol6.5 suggestion format
+                if isinstance(suggestions, dict):
+                    for word,sug in suggestions.items():
+                        spelling_suggestions = [item["word"] for item in sug['suggestion']]  #aggregate for future use in multi suggestion response
+                #Legacy Legacy & 6.4 handling
+                else:
+                    spelling_suggestions = []
+                    if isinstance(suggestions[1], dict):  #Solr6.4
+                        for item in suggestions:
+                            if isinstance(item, dict):
+                                spelling_suggestions += [subitem["word"] for subitem in item['suggestion']]
+                    else:  #Legacy Solr
+                        spelling_suggestions=suggestions[-1]
+
+                spelling_suggestion = spelling_suggestions[-1]  #Keep current method of returning single value
```

Comment:
Since we have this code in a few places, what do you think about declaring `spelling_suggestions` once, having the various branches populate it, and then having the check to return None or a single value at one place afterwards?